ambry-tools/src/main/java/com.github.ambry.tools/perf/rest/PerfRSC.java

```diff
@@ -0,0 +1,139 @@
+/**
+ * Copyright 2016 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.tools.perf.rest;
+
+import com.github.ambry.router.AsyncWritableChannel;
+import com.github.ambry.router.Callback;
+import com.github.ambry.router.FutureResult;
+import com.github.ambry.router.ReadableStreamChannel;
+import java.nio.ByteBuffer;
+import java.nio.channels.ClosedChannelException;
+import java.security.NoSuchAlgorithmException;
+import java.util.concurrent.Future;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+
+/**
+ * An implementation of {@link ReadableStreamChannel} that uses a byte array repeatedly to construct a stream of the
+ * specified size.
+ */
+class PerfRSC implements ReadableStreamChannel {
+  private final ByteBuffer bytes;
+  private final long streamSize;
+  private final AtomicBoolean readIntoInvoked = new AtomicBoolean(false);
+  private final AtomicBoolean callbackInvoked = new AtomicBoolean(false);
+  private final AtomicBoolean isOpen = new AtomicBoolean(true);
+
+  private volatile long streamed = 0;
+  private volatile Exception channelWriteException = null;
+
+  private volatile AsyncWritableChannel writeChannel = null;
+  private volatile FutureResult<Long> futureResult = new FutureResult<Long>();
+  private volatile Callback<Long> readIntoCallback = null;
+
+  /**
+   * Create an instance of PerfRSC.
+   * @param bytes the bytes that form a chunk that will be repeatedly returned until {@code streamSize} is reached.
+   * @param streamSize the desired total size of the stream.
+   */
+  public PerfRSC(byte[] bytes, long streamSize) {
+    if (streamSize < 0 || (streamSize > 0 && bytes.length < 1)) {
```

Comment:
What if `streamSize>=0 && bytes ==null`?

Seems only for non-POST request creation, a `null` byte array will be passed in with `streamSize=0`, however, this would not be an automatic assumption when just looking at this method.